lnwallet/rpcwallet/remote_signer_client.go

```diff
@@ -0,0 +1,977 @@
+				case <-returnedChan:
+				}
+
+			// An error occurred during the registration process.
+			case *walletrpc.RegistrationResponse_RegistrationError:
+				err := fmt.Errorf("registration error: %s",
+					rType.RegistrationError)
+
+				select {
+				case errChan <- err:
+				case <-returnedChan:
+				}
+			}
+
+			return
+
+		default:
+			err := fmt.Errorf("expected registration response, "+
+				"but got: %T", reqType)
+
+			select {
+			case errChan <- err:
+			case <-returnedChan:
+			}
+
+			return
+		}
+	}()
+
+	// Wait for the watch-only node to respond that it has accepted the
+	// signer has registered.
+	select {
+	case <-completeChan:
+		// TODO(viktor): This should verify that the signature in the
+		// complete message is valid.
+
+	case err := <-errChan:
+		return fmt.Errorf("watch-only node handshake error: %w", err)
+
+	case <-r.quit:
+		return ErrShuttingDown
+
+	case <-streamCtx.Done():
+		return streamCtx.Err()
+
+	case <-time.After(r.requestTimeout):
+		return errors.New("watch-only node handshake timeout")
+	}
+
+	return nil
+}
+
+// processSignRequests processes and responds to the sign requests that are
+// sent over the stream. The function will continuously run until the remote
+// signer client is either stopped or the stream errors.
+func (r *OutboundClient) processSignRequests(streamCtx context.Context) error {
+	var (
+		reqChan = make(chan *walletrpc.SignCoordinatorRequest)
+		errChan = make(chan error)
+	)
+
+	// We run the receive loop in a goroutine to ensure we can stop if the
+	// remote signer client is shutting down (i.e. the quit channel is
+	// closed). Closing the quit channel will make the processSignRequests
+	// function return, which will cancel the stream context, which in turn
+	// will stop the receive goroutine.
+	r.wg.Add(1)
+	go func() {
+		defer r.wg.Done()
+
+		for {
+			req, err := r.stream.Recv()
+			if err != nil {
+				wrappedErr := fmt.Errorf("error receiving "+
+					"request from watch-only node: %w", err)
+
+				// Send the error to the error channel, given
+				// that we're still listening on the channel.
+				select {
+				case errChan <- wrappedErr:
+				case <-streamCtx.Done():
+				case <-r.quit:
+				}
+
+				return
+			}
+
+			select {
+			case <-streamCtx.Done():
+				return
+
+			case <-r.quit:
+				return
+
+			case reqChan <- req:
+			}
+		}
+	}()
+
+	for {
+		log.Tracef("Waiting for a request from the watch-only node")
+
+		select {
+		case req := <-reqChan:
+			// Process the received request.
+			err := r.handleRequest(streamCtx, req)
+			if err != nil {
+				return err
+			}
+
+		case <-r.quit:
+			return ErrShuttingDown
+
+		case <-streamCtx.Done():
+			return streamCtx.Err()
+
+		case err := <-errChan:
+			return err
+		}
+	}
+}
+
+// handleRequest processes the received request from the watch-only node, and
+// sends the corresponding response back.
+func (r *OutboundClient) handleRequest(streamCtx context.Context,
```

Comment:
```
func (r *OutboundClient) handleRequest(ctx context.Context,
	req *walletrpc.SignCoordinatorRequest, stream StreamClient) error {

	log.Debugf("Processing a request from watch-only node of type: %T",
		req.GetSignRequestType())

	log.Tracef("Request content: %v", formatSignCoordinatorMsg(req))

	// Process the request.
	resp, err := r.process(ctx, req)
	if err != nil {
		errStr := "error processing the request in the remote " +
			"signer: " + err.Error()

		log.Errorf(errStr)

		// If we fail to process the request, we will send a SignerError
		// back to the watch-only node, indicating the nature of the
		// error.
		eType := &walletrpc.SignCoordinatorResponse_SignerError{
			SignerError: &walletrpc.SignerError{
				Error: errStr,
			},
		}

		resp = &signerResponse{
			RefRequestId:     req.GetRequestId(),
			SignResponseType: eType,
		}
	}

	// Send the response back to the watch-only node.
	err = stream.Send(resp)
	if err != nil {
		return fmt.Errorf("error sending response to watch-only "+
			"node: %w", err)
	}

	log.Tracef("Sent the following response to watch-only node: %v",
		formatSignCoordinatorMsg(resp))

	return nil
}
```

Reply:
Adding a comment to the things I couldn't change here either:

func (r *OutboundClient) handleRequest(ctx context.Context,
	req *walletrpc.SignCoordinatorRequest, stream StreamClient) error {

	log.Debugf("Processing a request from watch-only node of type: %T",
		req.GetSignRequestType())

	log.Tracef("Request content: %v", formatSignCoordinatorMsg(req))

	// Process the request.
	resp, err := r.process(ctx, req)
	if err != nil {
		errStr := "error processing the request in the remote " +
			"signer: " + err.Error()

		log.Errorf(errStr)

		// If we fail to process the request, we will send a SignerError
		// back to the watch-only node, indicating the nature of the
		// error.
		eType := &walletrpc.SignCoordinatorResponse_SignerError{
			SignerError: &walletrpc.SignerError{
				Error: errStr,
			},
		}

		resp = &signerResponse{
			RefRequestId:     req.GetRequestId(),
			SignResponseType: eType,
		}
	}

	// Send the response back to the watch-only node.
--------
	I think this comes with the same issue as in the handshake. We can't check if the send times out if we do it this way.
	I could implement a SendWithTimeout func on the stream though if you think that's a better solution.
-------
	err = stream.Send(resp)
	if err != nil {
		return fmt.Errorf("error sending response to watch-only "+
			"node: %w", err)
	}

	log.Tracef("Sent the following response to watch-only node: %v",
		formatSignCoordinatorMsg(resp))

	return nil
}